kep/624-disk-io-aware-scheduling/README.md

```diff
@@ -0,0 +1,207 @@
+
+The IO Driver, which is to be implemented by disk IO vendors, comprises three components.
+
+### IO Metrics Collector
+
+The IO Metrics Collector, which runs on each worker node, acts as an IO metric collector and analyzer. It watches the actual disk IO utilization of each pod, calculates the disk’s available IO capacity based on each workload’s real-time usage and characteristics using the IO Calculation Model and reports the node’s real-time available disk IO capacity to the aggregator when it is smaller than what is saved in the scheduler plugin’s cache. Since different disk vendors could have different ways of collecting metric, this component is outside the scope of the scheduler plugin’s implementation.
+
+### Aggregator
+
+The aggregator consolidates the IO metrics from each worker node, reports a consolidated list of real-time available disk IO capacity to the scheduler plugin. It is also responsible for converting each new pod’s disk IO BW request to a normalized value using the disk’s IO Calculation Model so as to match the disk’s normalized available IO capacity. Since the normalization function is disk type and vendor specific, this component is outside the scope of the scheduler plugin’s implementation.
+
+### IO Calculation Model
+
+The IO Calculation Model is responsible for converting the disk’s available IO capacity and each new pod’s IO request to normalized values. Since the normalization function is disk type and vendor specific, this component is outside the scope of the scheduler plugin’s implementation.
+
+### Disk IO Scheduler Plugin
+
+We leverage the K8s scheduler framework to add the disk IO scheduler plugin.
+When the scheduler plugin starts, it subscribes to the IO Driver with a long streaming connection. The IO Driver updates the scheduler plugin about each node disk’s normalized available IO capacity and the scheduler plugin stores the info in the plugin’s local cache. We thought about using CRD mechanism before. But going through the CRD route, it has two drawbacks: 
+1. In the pod scheduling flow, the disk IO scheduler plugin needs to pass the disk IO info specified in the annotation of the pod spec to the vendor-specific IO driver and get back a normalized disk IO BW value. Using CRD to support this bi-directional communication would involve creating two CRDs and two watches, which introduces long latency to the pod scheduling flow. 
+2. A node's disk IO available capacity would change dynamically upon the workloads' real-time IO access characteristics such as block size and IO read/write ratio. If we choose the CRD route, the real-time disk IO available capacity update will inject too much traffic into the API server and degrade the API server's performance. Using a direct communication channel between the disk IO scheduler plugin and the vendor-specific IO driver greatly helps to mitigate these two issues. 
```

Comment:
Using API to unify all components if the core of k8s ecosystem. We shouldn't arbitrarily plumb connections among any two components directly. Furthermore, treat scheduler as one component in the big picture, its functionality is also used by other ones like cluster autoscaler. If this design is plumbed into scheduler, it'd mean it needs to be carry over to those components as well.

> If we choose the CRD route, the real-time disk IO available capacity update will inject too much traffic into the API server and degrade the API server's performance.

Did you benchmark the API server's performance?

Reply:
Our original design is for each node to send the real-time disk IO telemetry info (such as real-time disk IO BW usage of each pod and real-time available disk IO capacity of each node) to the API server. However, we found out that it injects too much traffic into the API server which degrades the API server performance. Therefore we optimize it by adding an intermediate aggregator that performs telemetry traffic aggregation and batch transport and letting the scheduler plugin handle the interaction with the aggregator. Another consideration of the new design is that this telemetry info changes dynamically, there is no need to persist this info in the API server/ETCD, and therefore no need to go through the CRD route.

Reply:
Per scheuduling's perspective, the minimum needed info is the Node's available disk IO capacity, I'm skeptical that sort of into plus a reasonable sync internal would cause a downgrade on API server.

Reply:
The impact also depends on the number of nodes in a cluster, especially if it is a large cluster with thousands of nodes. We will re-evaluate the impact and see if it is OK to feed the telemetry info through the API server route.

Reply:
> Using API to unify all components if the core of k8s ecosystem. We shouldn't arbitrarily plumb connections among any two components directly. Furthermore, treat scheduler as one component in the big picture, its functionality is also used by other ones like cluster autoscaler. If this design is plumbed into scheduler, it'd mean it needs to be carry over to those components as well.
> 
> > If we choose the CRD route, the real-time disk IO available capacity update will inject too much traffic into the API server and degrade the API server's performance.
> 
> Did you benchmark the API server's performance?

Hi, Wei. We did benchmark API server's performance. The test is performed in k8s v1.25.  As scheduling accuracy depends on the node metric update frequency, our update interval is 1s. We measured the P99 latency and failed request rate when sending createPod requests to API Server along with concurrent CR updates to API server. Telling from our test result, when there are 1500 simultaneous node status updates within a one-second interval, the P99 latency will be over 0.5s and the rate of failed requests will exceed 10%. Do you think it is an expected behavior?

Reply:
@Huang-Wei Our test shows that the larger the cluster, the higher the failure rate. If the 10% failure rate for a cluster of 1,500 nodes is acceptable, we can change our design and feed the telemetry info via the API server route. I feel the failure rate is kind of high if we go for the API server route.

Reply:
I cannot speak for the upper bound of API Server/etcd, but updating interval to 1s sounds short. (ref: the default heartbeat interval is much higher: https://kubernetes.io/docs/reference/node/node-status/#heartbeats)

> we can change our design and feed the telemetry info via the API server route. I feel the failure rate is kind of high if we go for the API server route.

Generally I've seen similar solutions to impose delta resource changes to Node or CR objects when necessary, which serves as an input to make the resource aware by scheduler, and I believe this is the right approach to proceed with.